pkg/internal/util/cli/logger.go

```diff
@@ -20,58 +20,105 @@ import (
 	"bytes"
 	"fmt"
 	"io"
+	"runtime"
+	"strings"
 	"sync"
 
 	"sigs.k8s.io/kind/pkg/log"
 )
 
 // Logger is the kind cli's log.Logger implementation
 type Logger struct {
-	Verbosity log.Level
-	io.Writer
-	writeMu sync.Mutex
+	writer     io.Writer
+	writerMu   sync.Mutex
+	verbosity  log.Level
+	bufferPool *bufferPool
 }
 
 var _ log.Logger = &Logger{}
 
 // NewLogger returns a new Logger with the given verbosity
 func NewLogger(writer io.Writer, verbosity log.Level) *Logger {
 	return &Logger{
-		Verbosity: verbosity,
-		Writer:    writer,
+		verbosity:  verbosity,
+		writer:     writer,
+		bufferPool: newBufferPool(),
 	}
 }
 
-func (l *Logger) Write(p []byte) (n int, err error) {
-	// TODO: line oriented instead?
-	// For now we make a single per-message write call from the rest of the logger
-	// intentionally to effectively do this one level up
-	l.writeMu.Lock()
-	defer l.writeMu.Unlock()
-	return l.Writer.Write(p)
+// synchronized write to the inner writer
+func (l *Logger) write(p []byte) (n int, err error) {
+	l.writerMu.Lock()
+	defer l.writerMu.Unlock()
+	return l.writer.Write(p)
 }
 
-// TODO: prefix log lines with metadata (log level? timestamp?)
-
-func (l *Logger) print(message string) {
-	buf := bytes.NewBufferString(message)
+// writeBuffer writes buf with write, ensuring there is a trailing newline
+func (l *Logger) writeBuffer(buf *bytes.Buffer) {
+	// ensure trailing newline
 	if buf.Bytes()[buf.Len()-1] != '\n' {
 		buf.WriteByte('\n')
 	}
 	// TODO: should we handle this somehow??
 	// Who logs for the logger? 🤔
-	_, _ = l.Write(buf.Bytes())
+	_, _ = l.write(buf.Bytes())
```

Comment:
panic?

Reply:
That seems possibly excessive? 🤔

Reply:
just thinking in loud, what will be the problem if we can't write to the buffer?
should it be so important that we need to stop kind? or we can just move ahead simply ignoring it?

Reply:
writing to the buffer doesn't return an error because it can only panic when failing to allocat memory

Reply:
that is, your specific example is actually well documented to only return `nil` when implementing the writer API, and the only known failure mode is failure to acquire memory, which is a runtime error in go that isn't expected to be handled.

for writing to the output writer (currently only stderr anywhere that can instance this publicly) i don't think we can do much and i'm not sure panicing is ideal. I'll TODO cross checking with klog, as this is not new to this PR

Reply:
https://github.com/kubernetes/klog/blob/d39ecebdf3ba33f74b18becf8792d14bfb60576f/klog.go#L824

klog does not handle any write errors to output, so we're at least on par with the rest of the kubernetes ecosystem here